xcm/pallet-xcm/src/lib.rs

```diff
@@ -763,700 +957,968 @@ pub mod pallet {
+		Ok(())
+	}
+}
+
+pub struct LockTicket<T: Config> {
+	sovereign_account: T::AccountId,
+	amount: BalanceOf<T>,
+	unlocker: MultiLocation,
+	item_index: Option<usize>,
+}
+
+impl<T: Config> xcm_executor::traits::Enact for LockTicket<T> {
+	fn enact(self) -> Result<(), xcm_executor::traits::LockError> {
+		use xcm_executor::traits::LockError::UnexpectedState;
+		let mut locks = LockedFungibles::<T>::get(&self.sovereign_account).unwrap_or_default();
+		match self.item_index {
+			Some(index) => {
+				ensure!(locks.len() > index, UnexpectedState);
+				ensure!(locks[index].1.try_as::<_>() == Ok(&self.unlocker), UnexpectedState);
+				locks[index].0 = locks[index].0.max(self.amount);
+			},
+			None => {
+				locks
+					.try_push((self.amount, self.unlocker.clone().into()))
+					.map_err(|()| UnexpectedState)?;
+			},
 		}
+		LockedFungibles::<T>::insert(&self.sovereign_account, locks);
+		T::Currency::extend_lock(
+			*b"py/xcmlk",
+			&self.sovereign_account,
+			self.amount,
+			WithdrawReasons::all(),
+		);
+		Ok(())
+	}
+}
 
-		/// Note that a particular destination to whom we would like to send a message is unknown
-		/// and queue it for version discovery.
-		fn note_unknown_version(dest: &MultiLocation) {
-			log::trace!(
-				target: "xcm::pallet_xcm::note_unknown_version",
-				"XCM version is unknown for destination: {:?}",
-				dest,
-			);
-			let versioned_dest = VersionedMultiLocation::from(dest.clone());
-			VersionDiscoveryQueue::<T>::mutate(|q| {
-				if let Some(index) = q.iter().position(|i| &i.0 == &versioned_dest) {
-					// exists - just bump the count.
-					q[index].1.saturating_inc();
-				} else {
-					let _ = q.try_push((versioned_dest, 1));
+pub struct UnlockTicket<T: Config> {
+	sovereign_account: T::AccountId,
+	amount: BalanceOf<T>,
+	unlocker: MultiLocation,
+}
+
+impl<T: Config> xcm_executor::traits::Enact for UnlockTicket<T> {
+	fn enact(self) -> Result<(), xcm_executor::traits::LockError> {
+		use xcm_executor::traits::LockError::UnexpectedState;
+		let mut locks =
+			LockedFungibles::<T>::get(&self.sovereign_account).ok_or(UnexpectedState)?;
+		let mut maybe_remove_index = None;
+		let mut locked = BalanceOf::<T>::zero();
+		let mut found = false;
+		// We could just as well do with with an into_iter, filter_map and collect, however this way
+		// avoids making an allocation.
+		for (i, x) in locks.iter_mut().enumerate() {
+			if x.1.try_as::<_>().defensive() == Ok(&self.unlocker) {
+				x.0 = x.0.saturating_sub(self.amount);
+				if x.0.is_zero() {
+					maybe_remove_index = Some(i);
 				}
-			});
+				found = true;
+			}
+			locked = locked.max(x.0);
+		}
+		ensure!(found, UnexpectedState);
+		if let Some(remove_index) = maybe_remove_index {
+			locks.swap_remove(remove_index);
 		}
+		LockedFungibles::<T>::insert(&self.sovereign_account, locks);
+		let reasons = WithdrawReasons::all();
+		T::Currency::set_lock(*b"py/xcmlk", &self.sovereign_account, locked, reasons);
```

Comment:
I also tested the Lock and Unlock process, and looks to me as if `set_lock` is creating a new lock with the difference between what was locked before and the amount that is being unlocked. However, with pallet-balances implementation of set_lock 

https://github.com/paritytech/substrate/blob/7d233c2446b5a60662400a0a4bcfb78bb3b79ff7/frame/balances/src/lib.rs#L2133

when we are unlocking the full lock value, the unlock does not apply. Maybe we would have to call `remove_lock` in this case?